File: e2e/helpers/mailhog-client.ts
/**
 * MailHog 邮件捕获客户端。
 *
 * 在 E2E 测试中替代真实 SMTP，可重复捕获邮件并解析链接。
 * MailHog API: http://localhost:8025/api/v2/messages
 *
 * 用途（Phase 1）：
 * - 捕获邮箱验证邮件 + 邀请邮件
 * - 解析邀请链接 token（完成注册流程）
 *
 * Phase 3 扩展：Outlook Safe Links UA 模拟（见 safe-links-future-proof.spec.ts）
 */

const MAILHOG_API_URL = process.env.MAILHOG_API_URL ?? 'http://localhost:8025';

/** MailHog 邮件收件人格式 */
interface MailHogRecipient {
  Relays: null | string[];
  Mailbox: string;
  Domain: string;
  Params: string;
}

/** MailHog 邮件内容 */
interface MailHogContent {
  Headers: Record<string, string[]>;
  Body: string;
  Size: number;
  MIME: null | Record<string, unknown>;
}

/** MailHog 单封邮件结构 */
export interface MailHogMessage {
  ID: string;
  From: MailHogRecipient;
  To: MailHogRecipient[];
  Content: MailHogContent;
  Created: string;
  MIME: null | Record<string, unknown>;
  Raw: {
    From: string;
    To: string[];
    Data: string;
    Helo: string;
  };
}

/** MailHog 列表响应 */
interface MailHogListResponse {
  total: number;
  count: number;
  start: number;
  items: MailHogMessage[];
}

/**
 * 轮询 MailHog，直到找到发给指定邮箱的最新邮件。
 *
 * @param toEmail  收件人邮箱地址
 * @param timeout  最大等待毫秒数（默认 10000）
 * @returns        找到的邮件
 * @throws         超时未找到时抛出 Error
 */
export async function getLatestEmail(
  toEmail: string,
  timeout = 10_000,
): Promise<MailHogMessage> {
  const start = Date.now();

  while (Date.now() - start < timeout) {
    const r = await fetch(`${MAILHOG_API_URL}/api/v2/messages?limit=50`);

    if (!r.ok) {
      throw new Error(`MailHog API 返回 ${r.status}，请确认 MailHog 已启动`);
    }

    const data: MailHogListResponse = await r.json();
    const found = data.items.find((m) =>
      m.To.some(
        (t) => `${t.Mailbox}@${t.Domain}`.toLowerCase() === toEmail.toLowerCase(),
      ),
    );

    if (found) return found;

    await new Promise((resolve) => setTimeout(resolve, 300));
  }

  throw new Error(`在 ${timeout}ms 内未找到发给 ${toEmail} 的邮件`);
}

/**
 * 从邮件 HTML body 中提取包含指定字符串的第一个 href 链接。
 *
 * @param email        MailHog 邮件对象
 * @param hrefContains 链接中必须包含的子字符串（如 '/invite?token='）
 * @returns            完整的链接 URL
 * @throws             未找到时抛出 Error
 */
export function extractLink(email: MailHogMessage, hrefContains: string): string {
  const body = email.Content.Body;

  // 尝试匹配 href="..."
  const patterns = [
    new RegExp(`href="([^"]*${escapeRegex(hrefContains)}[^"]*)"`, 'i'),
    new RegExp(`href='([^']*${escapeRegex(hrefContains)}[^']*)'`, 'i'),
  ];

  for (const pattern of patterns) {
    const m = body.match(pattern);
    if (m?.[1]) return m[1];
  }

  // 纯文本链接（fallback）
  const textPattern = new RegExp(
    `(https?://[^\\s<>"']*${escapeRegex(hrefContains)}[^\\s<>"']*)`,
    'i',
  );
  const tm = body.match(textPattern);
  if (tm?.[1]) return tm[1];

  throw new Error(
    `在邮件中未找到包含 "${hrefContains}" 的链接\n邮件 Body 前 500 字符: ${body.slice(0, 500)}`,
  );
}

/**
 * 清空 MailHog 所有邮件（每个 spec 运行前调用以保证隔离）。
 */
export async function purgeAllEmails(): Promise<void> {
  await fetch(`${MAILHOG_API_URL}/api/v1/messages`, {
    method: 'DELETE',
  });
}

/**
 * 获取所有邮件列表。
 */
export async function getAllEmails(): Promise<MailHogMessage[]> {
  const r = await fetch(`${MAILHOG_API_URL}/api/v2/messages?limit=100`);
  if (!r.ok) return [];
  const data: MailHogListResponse = await r.json();
  return data.items;
}

/**
 * 等待 MailHog 接收到发给某邮箱的邮件，返回邮件中匹配的链接。
 * 组合函数：getLatestEmail + extractLink
 */
export async function waitForEmailAndExtractLink(
  toEmail: string,
  hrefContains: string,
  timeout = 15_000,
): Promise<string> {
  const email = await getLatestEmail(toEmail, timeout);
  return extractLink(email, hrefContains);
}

/** Phase 3 HITL 邮件解析后的结构 */
export interface HitlEmailParsed {
  /** 收件人邮箱 */
  to: string;
  /** 发件人邮箱 */
  from: string;
  /** subject 头 */
  subject: string;
  /** 完整 HTML body */
  html: string;
  /** 纯文本 fallback（如有） */
  text: string;
  /** 是否催办邮件（subject 含 [催办] 或 [升级] 前缀） */
  isReminder: boolean;
  /** 是否升级邮件（subject 含 [升级] 前缀） */
  isEscalation: boolean;
  /** 提取的 deeplink 列表 */
  deeplinks: Array<{
    /** action 名称（推断） */
    action: string;
    /** 完整 URL */
    url: string;
    /** JWT token 字符串 */
    token: string;
    /** JWT payload 中的 jti */
    jti: string;
  }>;
}

/**
 * 轮询 MailHog 直到找到发给 recipient 的最新 HITL 邮件，并解析 button + token。
 *
 * @param recipient 收件人邮箱
 * @param timeout   最大等待毫秒数（默认 15000）
 * @returns         结构化解析后的 HITL 邮件
 *
 * **使用约定**：
 * - 调用方应先 purgeAllEmails() 保证测试隔离
 * - 邮件 HTML 必须包含至少 1 个 `/hitl/page/<token>` 链接
 */
export async function getLatestHitlEmail(
  recipient: string,
  timeout = 15_000,
): Promise<HitlEmailParsed> {
  const email = await getLatestEmail(recipient, timeout);

  // 解析头部
  const subject = decodeMimeHeader(email.Content.Headers['Subject']?.[0] ?? '');
  const from = email.Content.Headers['From']?.[0] ?? '';
  const to = email.Content.Headers['To']?.[0] ?? '';

  // 获取完整 body
  const fullBody = email.Content.Body;

  // 提取 HTML 与 text part（MIME multipart 简单解析）
  const { html, text } = splitMimeBody(fullBody);

  // 解析 deeplinks
  const deeplinks = parseDeeplinks(html || fullBody);

  return {
    to,
    from,
    subject,
    html: html || fullBody,
    text,
    isReminder: subject.includes('[催办]'),
    isEscalation: subject.includes('[升级]'),
    deeplinks,
  };
}

/**
 * 从 HITL 邮件 HTML 中提取所有 token。
 *
 * 与 getLatestHitlEmail 互补：当只需要 token 列表（不需完整邮件信息）时使用。
 */
export function extractTokensFromEmail(html: string): Array<{
  action: string;
  jti: string;
  token: string;
}> {
  return parseDeeplinks(html).map(({ action, jti, token }) => ({ action, jti, token }));
}

/**
 * 解析邮件 HTML 中的 HITL deeplink button。
 *
 * 匹配两种格式：
 * 1. `<a href="...hitl/page/TOKEN...">通过</a>`（HTML button）
 * 2. `https://.../hitl/page/TOKEN`（plain text URL）
 */
function parseDeeplinks(body: string): Array<{
  action: string;
  url: string;
  token: string;
  jti: string;
}> {
  const result: Array<{
    action: string;
    url: string;
    token: string;
    jti: string;
  }> = [];
  const seenTokens = new Set<string>();

  // 1. HTML `<a href="...">label</a>` 模式
  const htmlPattern =
    /<a\s+[^>]*href=["']([^"']*\/hitl\/page\/([A-Za-z0-9._-]+))[^"']*["'][^>]*>([^<]*)<\/a>/gi;
  let m: RegExpExecArray | null;

  while ((m = htmlPattern.exec(body)) !== null) {
    const url = m[1];
    const token = m[2];
    const label = (m[3] ?? '').trim();
    if (seenTokens.has(token)) continue;
    seenTokens.add(token);
    result.push({
      action: inferAction(label, url),
      url,
      token,
      jti: extractJtiFromJwt(token),
    });
  }

  // 2. plain text URL（fallback —— 邮件 plain text 没有 HTML 包装）
  const textPattern = /(https?:\/\/[^\s<>"']*\/hitl\/page\/([A-Za-z0-9._-]+))/gi;
  while ((m = textPattern.exec(body)) !== null) {
    const url = m[1];
    const token = m[2];
    if (seenTokens.has(token)) continue;
    seenTokens.add(token);
    result.push({
      action: inferAction('', url),
      url,
      token,
      jti: extractJtiFromJwt(token),
    });
  }

  return result;
}

/** 按 button 文案 + URL 中的查询参数推断 action */
function inferAction(label: string, url: string): string {
  const lower = label.toLowerCase();
  if (lower.includes('通过') || lower.includes('approve')) return 'approve';
  if (lower.includes('提交') || lower.includes('submit')) return 'submit';
  if (lower.includes('退回') || lower.includes('return')) return 'return';
  if (lower.includes('拒绝') || lower.includes('reject')) return 'reject';

  // 从 URL 查询参数推断（如 ?action=submit）
  const urlMatch = url.match(/[?&]action=(\w+)/);
  if (urlMatch) return urlMatch[1];

  return 'unknown';
}

/** 从 JWT 提取 jti（不校验签名） */
function extractJtiFromJwt(token: string): string {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return '';
    const payloadBase64 = parts[1];
    const base64 = payloadBase64.replace(/-/g, '+').replace(/_/g, '/');
    const padded = base64 + '='.repeat((4 - (base64.length % 4)) % 4);
    const payloadJson = Buffer.from(padded, 'base64').toString('utf-8');
    const payload = JSON.parse(payloadJson) as { jti?: string };
    return payload.jti ?? '';
  } catch {
    return '';
  }
}

/** Decode quoted-printable / =?UTF-8?B?...?= 编码的 MIME header */
function decodeMimeHeader(raw: string): string {
  // RFC 2047 encoded-word: =?charset?B|Q?encoded-text?=
  return raw.replace(
    /=\?([^?]+)\?([BbQq])\?([^?]*)\?=/g,
    (_, _charset: string, enc: string, text: string) => {
      try {
        if (enc.toUpperCase() === 'B') {
          return Buffer.from(text, 'base64').toString('utf-8');
        }
        // Q encoding（quoted-printable）
        return text.replace(/_/g, ' ').replace(/=([0-9A-F]{2})/g, (_, hex: string) =>
          String.fromCharCode(parseInt(hex, 16)),
        );
      } catch {
        return text;
      }
    },
  );
}

/** Split multipart MIME body into HTML and text parts */
function splitMimeBody(body: string): { html: string; text: string } {
  // 单一格式：body 不含 boundary → 直接判断 HTML/text
  if (!body.includes('Content-Type:') && !body.includes('boundary=')) {
    if (body.trimStart().startsWith('<') || body.includes('<html')) {
      return { html: body, text: '' };
    }
    return { html: '', text: body };
  }

  // multipart 简单切分
  let html = '';
  let text = '';

  // 找 text/html part
  const htmlMatch = body.match(
    /Content-Type:\s*text\/html[^\n]*\n(?:[^\n]*\n)*?\n([\s\S]*?)(?:\n--[a-zA-Z0-9_-]+|\n\.|\Z)/i,
  );
  if (htmlMatch) {
    html = decodeMimeBody(htmlMatch[1]);
  }

  // 找 text/plain part
  const textMatch = body.match(
    /Content-Type:\s*text\/plain[^\n]*\n(?:[^\n]*\n)*?\n([\s\S]*?)(?:\n--[a-zA-Z0-9_-]+|\n\.|\Z)/i,
  );
  if (textMatch) {
    text = decodeMimeBody(textMatch[1]);
  }

  // fallback：什么都没匹配，把整 body 当 html
  if (!html && !text) {
    html = body;
  }

  return { html, text };
}

/** 解码 quoted-printable / base64 编码的 MIME body */
function decodeMimeBody(raw: string): string {
  // quoted-printable: =XX → byte, =\n → 续行
  if (raw.includes('=')) {
    try {
      const decoded = raw
        .replace(/=\n/g, '')
        .replace(/=\r\n/g, '')
        .replace(/=([0-9A-F]{2})/gi, (_, hex: string) =>
          String.fromCharCode(parseInt(hex, 16)),
        );
      // 检查是否为 utf-8 字节序列
      return Buffer.from(decoded, 'binary').toString('utf-8');
    } catch {
      return raw;
    }
  }
  return raw;
}

/** 转义正则特殊字符 */
function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

File: e2e_v2/helpers/mailhog_client.py
"""MailHog 邮件捕获客户端（Python port 自 e2e/helpers/mailhog-client.ts）。

MailHog API: http://localhost:8025/api/v2/messages

提供：
- get_latest_email(to_email, timeout): 轮询找指定收件人邮件
- get_emails_for(to_email, timeout, min_count): 找 N 封发给同一人的邮件（chain mode）
- purge_all(): 清空所有邮件（spec 隔离）
- parse_hitl_deeplinks(html): 解析 HITL 邮件中的 4 按钮 deeplink
- extract_jti_from_token(token): 不校验签名提取 jti 字段
"""
from __future__ import annotations

import asyncio
import base64
import json
import os
import re
import time
from dataclasses import dataclass
from typing import Any

import httpx


MAILHOG_API_URL = os.environ.get("MAILHOG_API_URL", "http://localhost:8025")


# ── 数据结构 ────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class HitlDeeplink:
    """HITL 邮件按钮 deeplink（不可变快照）。"""

    action: str  # approve / return / reject / submit / unknown
    url: str
    token: str  # JWT 字符串
    jti: str


@dataclass(frozen=True)
class HitlEmailParsed:
    """解析后的 HITL 邮件结构。"""

    to: str
    subject: str
    html: str  # 主体 HTML
    text: str  # 纯文本 fallback
    is_reminder: bool  # subject 含 [催办]
    is_escalation: bool  # subject 含 [升级]
    is_supplement: bool  # subject 含 [终止] / [已被处理]
    deeplinks: tuple[HitlDeeplink, ...]


# ── HTTP fetch helpers ──────────────────────────────────────────────────


async def _fetch_messages(client: httpx.AsyncClient, limit: int = 50) -> list[dict]:
    """从 MailHog API 抓邮件列表。"""
    r = await client.get(f"{MAILHOG_API_URL}/api/v2/messages?limit={limit}")
    r.raise_for_status()
    return r.json().get("items", [])


def _msg_to_addresses(msg: dict) -> list[str]:
    """提取邮件的所有收件人 email（小写）。"""
    return [
        f"{t.get('Mailbox','')}@{t.get('Domain','')}".lower()
        for t in msg.get("To", [])
    ]


# ── 公共 API ────────────────────────────────────────────────────────────


async def purge_all(client: httpx.AsyncClient | None = None) -> None:
    """清空 MailHog 所有邮件。"""
    own = client is None
    if own:
        client = httpx.AsyncClient()
    try:
        await client.delete(f"{MAILHOG_API_URL}/api/v1/messages")
    finally:
        if own:
            await client.aclose()


async def get_latest_email(
    to_email: str,
    timeout: float = 10.0,
    client: httpx.AsyncClient | None = None,
) -> dict:
    """轮询找发给 to_email 的最新邮件（最近一封）。

    Raises:
        TimeoutError: timeout 秒内没找到匹配邮件
    """
    own = client is None
    if own:
        client = httpx.AsyncClient()
    try:
        deadline = time.time() + timeout
        target = to_email.lower()
        while time.time() < deadline:
            items = await _fetch_messages(client)
            for m in items:
                if target in _msg_to_addresses(m):
                    return m
            await asyncio.sleep(0.3)
        raise TimeoutError(f"{timeout}s 内未找到发给 {to_email} 的邮件")
    finally:
        if own:
            await client.aclose()


async def get_emails_for(
    to_email: str,
    min_count: int = 1,
    timeout: float = 15.0,
    client: httpx.AsyncClient | None = None,
) -> list[dict]:
    """轮询直到至少 min_count 封邮件发给 to_email（chain mode 验证 N 人同时收到）。

    返回所有匹配邮件（按 mailhog 顺序，最新在前）。
    """
    own = client is None
    if own:
        client = httpx.AsyncClient()
    try:
        deadline = time.time() + timeout
        target = to_email.lower()
        while time.time() < deadline:
            items = await _fetch_messages(client)
            matches = [m for m in items if target in _msg_to_addresses(m)]
            if len(matches) >= min_count:
                return matches
            await asyncio.sleep(0.3)
        raise TimeoutError(
            f"{timeout}s 内未找到 {min_count} 封发给 {to_email} 的邮件"
        )
    finally:
        if own:
            await client.aclose()


async def get_all_emails(client: httpx.AsyncClient | None = None) -> list[dict]:
    """获取 MailHog 所有邮件列表。"""
    own = client is None
    if own:
        client = httpx.AsyncClient()
    try:
        return await _fetch_messages(client, limit=200)
    finally:
        if own:
            await client.aclose()


# ── MIME body 切分 ──────────────────────────────────────────────────────


_QUOTED_PRINTABLE_BYTE = re.compile(rb"=([0-9A-Fa-f]{2})")


def _decode_quoted_printable(raw: str) -> str:
    """quoted-printable → utf-8。"""
    no_soft_break = raw.replace("=\n", "").replace("=\r\n", "")
    as_bytes = no_soft_break.encode("latin-1", errors="ignore")
    decoded_bytes = _QUOTED_PRINTABLE_BYTE.sub(
        lambda m: bytes.fromhex(m.group(1).decode()), as_bytes
    )
    return decoded_bytes.decode("utf-8", errors="replace")


def _decode_mime_header(raw: str) -> str:
    """=?UTF-8?B|Q?...?= 编码的 MIME header 解码（RFC 2047）。"""

    def _decode(match: re.Match[str]) -> str:
        enc = match.group(2).upper()
        text = match.group(3)
        try:
            if enc == "B":
                return base64.b64decode(text).decode("utf-8", errors="replace")
            return _decode_quoted_printable(text.replace("_", " "))
        except Exception:
            return text

    return re.sub(r"=\?([^?]+)\?([BbQq])\?([^?]*)\?=", _decode, raw)


def _split_mime_body(body: str) -> tuple[str, str]:
    """切分 multipart MIME body → (html, text)。"""
    if "Content-Type:" not in body and "boundary=" not in body:
        if body.lstrip().startswith("<") or "<html" in body:
            return body, ""
        return "", body

    html, text = "", ""
    html_match = re.search(
        r"Content-Type:\s*text/html[^\n]*\n(?:[^\n]*\n)*?\n([\s\S]*?)(?:\n--[a-zA-Z0-9_-]+|\n\.|\Z)",
        body,
        re.IGNORECASE,
    )
    if html_match:
        html = _decode_quoted_printable(html_match.group(1))

    text_match = re.search(
        r"Content-Type:\s*text/plain[^\n]*\n(?:[^\n]*\n)*?\n([\s\S]*?)(?:\n--[a-zA-Z0-9_-]+|\n\.|\Z)",
        body,
        re.IGNORECASE,
    )
    if text_match:
        text = _decode_quoted_printable(text_match.group(1))

    if not html and not text:
        html = body

    return html, text


# ── HITL 邮件解析 ───────────────────────────────────────────────────────


def extract_jti_from_jwt(token: str) -> str:
    """从 JWT token 解码 jti（不校验签名，仅 E2E 断言用）。"""
    try:
        parts = token.split(".")
        if len(parts) != 3:
            return ""
        payload_b64 = parts[1]
        # base64url → base64 + padding
        base64_str = payload_b64.replace("-", "+").replace("_", "/")
        padded = base64_str + "=" * ((4 - len(base64_str) % 4) % 4)
        payload_json = base64.b64decode(padded).decode("utf-8", errors="replace")
        payload = json.loads(payload_json)
        return str(payload.get("jti", ""))
    except Exception:
        return ""


def _infer_action(label: str, url: str) -> str:
    """按 button 文案 + URL fragment 推断 action。"""
    lower = label.lower()
    if "通过" in label or "approve" in lower:
        return "approve"
    if "提交" in label or "submit" in lower:
        return "submit"
    if "退回" in label or "return" in lower:
        return "return"
    if "拒绝" in label or "reject" in lower:
        return "reject"

    m = re.search(r"[?&]action=(\w+)", url)
    if m:
        return m.group(1)
    return "unknown"


_HITL_HTML_LINK = re.compile(
    r"""<a\s+[^>]*href=["']([^"']*/hitl/page/([A-Za-z0-9._-]+))[^"']*["'][^>]*>([^<]*)</a>""",
    re.IGNORECASE,
)
_HITL_TEXT_LINK = re.compile(
    r"""(https?://[^\s<>"']*/hitl/page/([A-Za-z0-9._-]+))""",
    re.IGNORECASE,
)


def parse_hitl_deeplinks(body: str) -> tuple[HitlDeeplink, ...]:
    """从邮件 HTML / plain text 中解析所有 /hitl/page/<token> 按钮。

    去重：同 token 仅记一次。
    """
    out: list[HitlDeeplink] = []
    seen: set[str] = set()

    for m in _HITL_HTML_LINK.finditer(body):
        url, token, label = m.group(1), m.group(2), m.group(3).strip()
        if token in seen:
            continue
        seen.add(token)
        out.append(
            HitlDeeplink(
                action=_infer_action(label, url),
                url=url,
                token=token,
                jti=extract_jti_from_jwt(token),
            )
        )

    for m in _HITL_TEXT_LINK.finditer(body):
        url, token = m.group(1), m.group(2)
        if token in seen:
            continue
        seen.add(token)
        out.append(
            HitlDeeplink(
                action=_infer_action("", url),
                url=url,
                token=token,
                jti=extract_jti_from_jwt(token),
            )
        )

    return tuple(out)


def parse_hitl_email(msg: dict) -> HitlEmailParsed:
    """把 MailHog message dict 解析成 HitlEmailParsed 结构。"""
    headers = msg.get("Content", {}).get("Headers", {})
    subject_raw = (headers.get("Subject", [""]) or [""])[0]
    subject = _decode_mime_header(subject_raw)
    to_raw = (headers.get("To", [""]) or [""])[0]
    to_addr = _decode_mime_header(to_raw)

    body = msg.get("Content", {}).get("Body", "")
    html, text = _split_mime_body(body)
    main_body = html or body

    return HitlEmailParsed(
        to=to_addr,
        subject=subject,
        html=html or body,
        text=text,
        is_reminder="[催办]" in subject,
        is_escalation="[升级]" in subject,
        is_supplement="[终止]" in subject or "已被" in subject,
        deeplinks=parse_hitl_deeplinks(main_body),
    )


async def get_latest_hitl_email(
    to_email: str,
    timeout: float = 15.0,
    client: httpx.AsyncClient | None = None,
) -> HitlEmailParsed:
    """组合：等邮件 → 解析为 HitlEmailParsed。"""
    raw = await get_latest_email(to_email, timeout=timeout, client=client)
    return parse_hitl_email(raw)
